##  Programming for Data Analytics Simulation Project 2020 
#### Module: Programming for Data Analytics
#### Student Name: John Kavanagh
#### Lecturer: Brian McGinley 

___

### Problem Statement 

For this project you must create a data set by simulating a real-world phenomenon of your choosing. You may pick any phenomenon you wish – you might pick one that is of interest to you in your personal or professional life. Then, rather than collect data
related to the phenomenon, you should model and synthesise such data using Python. We suggest you use the numpy.random package for this purpose.

Specifically, in this project you should:  
• Choose a real-world phenomenon that can be measured and for which you could collect at least one-hundred data points across at least four different variables.  
• Investigate the types of variables involved, their likely distributions, and their relationships with each other.  
• Synthesise/simulate a data set as closely matching their properties as possible.  
• Detail your research and implement the simulation in a Jupyter notebook – the data set itself can simply be displayed in an output cell within the notebook.  


Note that this project is about simulation – you must synthesise a data set. Some students may already have some real-world data sets in their own files. It is okay to base your synthesised data set on these should you wish (please reference it if you do), but the main task in this project is to create a synthesised data set.

### Example Guidance - can this be deleted??

As a lecturer I might pick the real-world phenomenon of the performance of students studying a ten-credit module. After some research, I decide that the most interesting variable related to this is the mark a student receives in the module - this is going to be one of my variables (grade).  
Upon investigation of the problem, I find that the number of hours on average a student studies per week (hours), the number of times they log onto Moodle in the first three weeks of term (logins), and their previous level of degree qualification (qual) are closely related to grade. The hours and grade variables will be non-negative real number with two decimal places, logins will be a non-zero integer and qual will be a categorical variable with four possible values: 
- none
- bachelors
- masters, or
- phd  

After some online research, I find that full-time post-graduate students study on average four hours per week with a standard deviation of a quarter of an hour and that a normal distribution is an acceptable model of such a variable. Likewise, I investigate the other four variables, and I also look at the relationships between the variables. I devise an algorithm (or method) to generate such a data set, simulating values of the four variables for two-hundred students. I detail all this work in my notebook, and then I add some code in to generate a data set with those properties. 

___

#### Software Requirements

There are a number of software packages htat will need to be installed throughout this exercise. 
Such as: 

- Anaconda
- Jupyter Notebook
- Pandas
- MatPlotLib
- Numpy
- Seaborn

___

### Project Overview    
#### Introduction  
#### Source of DataSet  
##### Analysis of Documentation
##### Analysis of DataSet  
###### DisPlot  
###### CatPlot  

#### Introduction to Numpy.Random  
##### Data Creation  
#####  
#####  
#####  


#### Introduction - Complete rewrite required
 
As part of the ressearch into this exercise, I thoughtn it best that we started off with an exisitng data set that was relevant to the information that i wanted to simulate, in order to provide some structure, guidance as well as some fallback information on the topic. 
The area that I wanted to cover is in relation to health. Obviously, one of the unavoidable aspects to life in 2020 hasm been COVID-19, as everyone is painfully aware. 
HOweverm, one of the aspects that appears to be examined is that of the relationshipn between hte NPHET committee and hte Government of Ireland. 

From the outset, there was a direct line of communication between the two boides, but as time went on, and the perview of the governments view fo teh world expeanded, there were layers of control put in place to provide distance. 

Instead of just public health, there were another numbers of factors to be catered for, such as The economy, mental health etc..

However, if we take the view of what the remit ofthe delivery of NPHET is for, in the strictest sense of the world, they are responsible for the public health. In Ireland public health has been a contenitious topic over decades. 
There has been a continuous rise in the amount of money required to work the system, with a lot of high profile costs being reported on. 


If we review the concept of advise and what becomes the recommended advise and or lae fro teh point of view of the governemnt, we are left to believe that there is no way we can have both. that we can't have a full lockdown, duw to the cpacity of hospitals as well as levels of training of ICU staff and nurses to cater for the re-assigned workload that OCVID has caused. 

Ireladn has suffered from a fundamental issue of how to manage it's health system for a long time. 


On a societal level, there are a number of changes that can be reviewed to reduce the level of activity that is burdened our system. One of the simple steps to ne taken, is to potentially review the general health of hte nation. 


A phenomenon, according to [Cambridge Dictionary](https://dictionary.cambridge.org/dictionary/english/phenomenon) is "something that exists and that can be seen, felt, tested etc.". 

___

### Statistical Society of Canada Dataset Overview
Given that we are looking to simulate infromation on the general area of health, it's import to be familiar with the structure of the information, as well as the headings that are reported upon. 

The first step in demonstrating this research was to review a site that held a large number of datasets across a number of disciplines. Having reviewed some initial datasets on [Vincent Arel-Bundock](https://vincentarelbundock.github.io/Rdatasets/datasets.html), there was a simple dataset with a small number of attributes that could be used as the basis for examination.   

This dataset, containing samples from 500 adults, as detailed in the documentation on [Vincent Ares-Bundock](https://vincentarelbundock.github.io/Rdatasets/doc/Stat2Data/Blood1.html) holds information on the Blood Pressure recorded as well as other factors such as smoking and whether the individual is classified as overweight. 

As part of the review, we are going to examine the data set for a relationship between blood pressure, smoking and a persons weight. 

Subsequent to this report, there was additional research required, in order to effectively ensure the report was being interpreted correctly, the extended report, as published on [Statistical Society of Canada](https://ssc.ca/en/case-studies-2003-annual-meeting) 

#### Analysis of Documentation

Although limited, there are a number of pieces of information on the [Vincent Arel-Bundock](https://vincentarelbundock.github.io/Rdatasets/doc/Stat2Data/Blood1.html) site that convey greater breadth of information on the data.

Smoking, for instance is captured in the affirmative of Yes. As per the Boolean expression, where there are two outcomes, categorised as either Zero or One, smokers are recorded as One.  

In relation to the other factor, the weight of an individual, there are three categories that have been created:  
1. Normal 
2. Overweight 
3. Obese  

Lastly, we have the recorded values of each individuals SystolicBP, or Blood Pressure as it is known. In order to ahve a more informed opinion on the range of what a healthy recording for blood pressure is, please review the following site, [IrishHeart.ie](https://irishheart.ie/your-health/ways-to-live-better/blood-pressure/)

#### Analysis on DataSet

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv('https://vincentarelbundock.github.io/Rdatasets/csv/Stat2Data/Blood1.csv')
df

Unnamed: 0  SystolicBP  Smoke  Overwt
0             1         133      0       2
1             2         115      1       0
2             3         140      1       1
3             4         132      0       2
4             5         133      0       1
..          ...         ...    ...     ...
495         496         146      0       0
496         497         181      1       1
497         498         145      1       2
498         499         180      1       2
499         500         174      1       2

[500 rows x 4 columns]

The package has successfully taken the URL and produced a tabular view of the information, with the headings that we expected to see. As you can see along the left hand side of teh table, pandas has added in a row for indexing the data.  

In [9]:
df.loc('Smoke')

ValueError: No axis named Smoke for object type DataFrame

In [4]:
df.describe()
# this will allow us to view some high-level information from the set, such as mean and the standard deviation.

Unnamed: 0  SystolicBP       Smoke      Overwt
count  500.000000  500.000000  500.000000  500.000000
mean   250.500000  144.952000    0.532000    1.034000
std    144.481833   27.994948    0.499475    0.884539
min      1.000000   67.000000    0.000000    0.000000
25%    125.750000  130.000000    0.000000    0.000000
50%    250.500000  140.500000    1.000000    1.000000
75%    375.250000  162.250000    1.000000    2.000000
max    500.000000  224.000000    1.000000    2.000000

In [5]:
# To effectively review the data, given that there are so few variables and the relationship is transparent at first sight,
# we import the Seaborn package.

In [ ]:
import seaborn as sns

#### DisPlot

In [ ]:
sns.displot(data=df, x="Overwt", y="SystolicBP", hue="Smoke")

Given that we know that three of the headings can create a view of the information (Unnamed not being an informative ingredient), we can create a plot that maps weight against blood pressure, and overlay on this information, those who have indicated that they smoke. 

We can see from the plot that: 
- If you do not smoke, you will not be categorised as overweight
- You're not going to siffer from a higher than normal blood pressure. 

Some of the information displayed isn't instantly readable, so we can try manipulate the information with a different variable.

Outstandings Questions for Review:  
- Identification of smoker: Does 0 mean they smoke  
- Overweight classification: we appear to have three bins that the participants have been categorised into  

  

___

IN oredr to introduce greater depth of understanding, I have chosen to introduce the catplot function. The catploty funciton allows for better visualisation given the type of data on display. As we work our way up the Y Axes, we can seee again that we have SystolicBP which represents Blood Pressure. 

    For more information inrelation to Blood Pressure, please review hte following site:
    https://irishheart.ie/your-health/ways-to-live-better/blood-pressure/
    
The important thing to note with this information is:
 - The normal level of blood pressure is usually about 120 (systolic) over 80 (diastolic)
 - If your blood pressure is 140 over 90 or higher (or 140 over 80 if you have diabetes) you should discuss this reading with your doctor.
 
We can now determine those who are unhealthy

#### CatPlot

In [ ]:
sns.catplot(data=df, x="Overwt", y="SystolicBP", hue="Smoke", kind="swarm")

We can also run a view that allows us to split the axis based on the values of the axis.  
In order to demonstrate this, I have chosen the smoker quesiton, as it has a Yes/No response. 

In [ ]:
sns.catplot(data=df, x="Smoke", y="SystolicBP", hue="Overwt", kind="violin")

Despite reviewing the data from a numbe rof different variable within the package, we can not definitively determine whether what the information relates to on the outstanding questions. Having reviewed the material that was supplementary to the dataset, there appears to be a larger dataset from which this was created. 


From reviewing this wider dataset, we can determine that: 

Smoking has been code to align to  1 as it is a Boolean expression.
Overwt = This classifcation has been stratified into 3 chunks, normal, overweight, obese.

In [ ]:
import matplotlib as plt

In [ ]:
fd = pd.read_csv("https://ssc.ca/sites/default/files/archive/documents/case_studies/2003/documents/datafile.dat")

In [ ]:
sns.jointplot(data=fd, x="age", y="weight",  hue="gender")

in order to form a wider relationship 

In [ ]:
sns.catplot(data=fd, x="smoke", y="sbp", hue="overwt", kind="violin")

____

#### Overview of Numpy.Random


In order to effectively simulate the data required to emulate the information as it is provided above, as a template for further testing, we must first understand the operational overview of the numpy.random package. 

Comprising of a generator and a bitgenerator, there is a melding of sequence creation and different statistical distributuions applied against to effectively generate random numbers, as per [Numpy](https://numpy.org/doc/stable/reference/random/index.html?highlight=numpy%20random#module-numpy.random). While the remit of this exercise is not to walk tnrough all components of the package, a high-level understanding will be displayed in code blocks and implementaion to better understand the material.  



In [21]:
import numpy as np
from numpy.random import default_rng
rng = default_rng()

rng

In [7]:
# In the output above we have the generator that numpy.random operates from. 

In [10]:
sq1 = np.random.SeedSequence()
sq1.entropy

319345192536006037530720654921075243083

In [12]:
from numpy.random import Generator, PCG64
rg = Generator(PCG64(12345))
rg.standard_normal()

-1.4238250364546312

In [15]:
from numpy.random import Generator, PCG64
rg = Generator(PCG64(12345))
rg.standard_normal()

-1.4238250364546312

If you are interested in the PCG (Permuted Congrunetila Generator-64bit), please click the following link:[pcg-random.org](https://www.pcg-random.org/)

Once we have stood up our PCG, and we are looking to create some data, it is important to note that the Numpy.raandom package has three core areas that can be leveraged against. 


- Simple Random Data 
- Permutations
- Distributions

At a high-level, ther are four sub-components that are part of Simple Random Data layer. 
They are: 
- Integer, Random, Choice, Bytes   

Permutations, as the name may suggesto those that are familiar to it, has a narrower scope, facilitating:
- Shuffle & Permutation  

Distributions, on the other hand, has a much wider span, with more than 35 distributions that can be accessed, such as 
- Power, Multinomial, Triangular & Wald  



#### Random Data Creation

In [33]:
rng.random()

0.0523101535315329

___

___
Bibliography


Seaborn Official site, retrieved on 30-12-2020
https://seaborn.pydata.org/introduction.html  

Pandas - Read a CSV File, retrieved on 30-12-2020
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html  

Vincent Arel-Bundock GitHub, retrieved on 30-12-2020 
https://vincentarelbundock.github.io/Rdatasets/datasets.html  

Vincent Arel-Bundock GitHub, retrieved on 30-12-2020
https://vincentarelbundock.github.io/Rdatasets/doc/Stat2Data/Blood1.html    

Statistical Society of Canada, retrieved on 30-12-2020 
https://ssc.ca/en/case-studies-2003-annual-meeting    

Cambridge Dictionary, retrieved on 30-12-2020
https://dictionary.cambridge.org/dictionary/english/phenomenon

Numpy.org, retrieved on 30-12-2020 
https://numpy.org/doc/stable/reference/random/index.html?highlight=numpy%20random#module-numpy.random  

PCG, A Family of Better Random Number Generators, retrieved on 30-12-2020 
https://www.pcg-random.org/

numpy.random.Generator.random, retrieved on 30-12-2020
https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.random.html#numpy.random.Generator.random

Permuted Congruential Generator (64-bit, PCG64), retrieved on 30-12-2020
https://numpy.org/doc/stable/reference/random/bit_generators/pcg64.html#numpy.random.PCG64


numpy.random.SeedSequence, retrieved on 30-12-2020
https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence